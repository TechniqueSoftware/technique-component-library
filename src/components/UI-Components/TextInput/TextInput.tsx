import * as React from 'react';
import FormControl, { FormControlProps } from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';
import InputAdornment from '@material-ui/core/InputAdornment';
import withStyles from '@material-ui/styles/withStyles';
import { Theme } from '@material-ui/core';
import { colors } from '../../themes/clubOS';
import { createStyles } from '@material-ui/styles';

const styles = (theme: Theme) => createStyles({
  root: {
    fontSize: theme.typography.body2.fontSize,
    display: 'flex',
    alignItems: 'center',
    background: theme.palette.common.white
  },
  input: {
    padding: theme.spacing(2),
    fontSize: theme.typography.body2.fontSize
  },
  inputNoLeftPad: {
    paddingLeft: 0
  },
  startIcon: {
    color: colors.icon.default(theme),
    paddingLeft: theme.spacing(2),
  }
});

/**
 * From T pick all properties except the set of properties K
 */
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export interface TextInputProps extends Omit<FormControlProps, 'ref'> {
  'aria-label': string;
  value: string;
  placeholder?: string;
  classes?: {
    root: string,
    startIcon: string,
    input: string,
    inputNoLeftPad: string
  };
  startAdornmentIcon?: React.ReactElement;
  autoFocus?: boolean;
}

export class TextInput extends React.PureComponent<TextInputProps> {

  renderStartAdornment() {
    const { startAdornmentIcon, classes } = this.props;
    if (!startAdornmentIcon) return null;

    return (<InputAdornment position="start" classes={{ root: classes.startIcon }}>
      {startAdornmentIcon}
    </InputAdornment>);
  }

  render() {
    const {
      'aria-label': ariaLabel,
      classes,
      startAdornmentIcon,
      placeholder,
      autoFocus = false,
      ...rest
    } = this.props;
    const inputClasses = [
      classes.input,
      startAdornmentIcon && classes.inputNoLeftPad,
    ].filter(Boolean).join(' ');
    return (
      <FormControl fullWidth={true} classes={{ root: classes.root }} {...rest}>
        <Input
          inputProps={{
            'aria-label': ariaLabel
          }}
          placeholder={placeholder}
          classes={{ input: inputClasses }}
          fullWidth={true}
          autoFocus={autoFocus}
          startAdornment={this.renderStartAdornment()}
        />
      </FormControl>
    );
  }
}

export default withStyles(styles)(TextInput);
