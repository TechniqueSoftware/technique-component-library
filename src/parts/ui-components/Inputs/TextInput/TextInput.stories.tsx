import * as React from 'react';
import { createCUICStoriesOf } from '../../../../components/UI-Components/utils/utils';
import { MODULE_NAME } from '../../../../components/UI-Components/constants';
import clubOsTheme from '../../../../components/UI-Components/storyThemes';
import { TextInput, TextInputProps } from '../../../..';
import { boolean, number, text } from '@storybook/addon-knobs';
import Search from '@material-ui/icons/Search';
import ThemeProvider from '@material-ui/styles/ThemeProvider';

type TextInputStoryProps = {
  containerWidth: number;
  placeholder: string;
  withAdornment: boolean;
};

type TextInputStoryState = { text: string };

class TextInputStory extends React.PureComponent<TextInputStoryProps, TextInputStoryState> {

  constructor(props: TextInputStoryProps) {
    super(props);
    this.state = { text: '' };
    this.handleTextChange = this.handleTextChange.bind(this);
  }

  handleTextChange(event: any) {
    this.setState({ text: event.target.value });
  }

  render() {
    const { containerWidth, placeholder, withAdornment } = this.props;
    const textInputProps: TextInputProps = {
      placeholder,
      'aria-label': placeholder,
      value: this.state.text,
      onChange: this.handleTextChange
    };
    if (withAdornment) {
      textInputProps.startAdornmentIcon = <Search />;
    }
    return (
      <div style={{ width: containerWidth }}>
        <ThemeProvider theme={clubOsTheme}>
          <TextInput {...textInputProps} />
        </ThemeProvider>
      </div>
    );
  }

}

createCUICStoriesOf(MODULE_NAME.INPUTS, module)
  .addCUICStory({
    render: () => {
      const placeholder = text('Placeholder text', 'Search');
      const containerWidth = number('Container width', 400);
      const withAddornment = boolean('With addornment icon', true);
      return (
        <TextInputStory
          placeholder={placeholder}
          containerWidth={containerWidth}
          withAdornment={withAddornment} />);
    },
    name: 'TextInput',
    notes: require('./notes.md').default
  });
