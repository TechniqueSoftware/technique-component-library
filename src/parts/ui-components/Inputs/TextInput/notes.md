# TextInput
This component wraps Material-UI's [FormControl](https://material-ui.com/api/form-control/) and [Input](https://material-ui.com/api/input/) 
components and adds custom stylings. 

## Usage
This component was created initially to be used in the `MultiSelectFieldGroup` component as a way to provide search.

## JIRA
Related jira ticket: https://clubos.atlassian.net/browse/CR-579
