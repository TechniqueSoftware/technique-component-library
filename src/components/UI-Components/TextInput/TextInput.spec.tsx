import { shallow } from 'enzyme';
import * as React from 'react';
import { TextInputProps, TextInput } from './TextInput';
import Search from '@material-ui/icons/Search';

describe('TextInput', () => {

  const baseProps: TextInputProps = {
    'aria-label': 'mockAriaLabel',
    value: 'mockValue',
    classes: {
      root: 'root',
      input: 'input',
      startIcon: 'startIcon',
      inputNoLeftPad: 'inputNoLeftPad'
    },
    onChange: jest.fn(),
    placeholder: 'mockPlaceHolder',
  };

  it('should render with minimal props', () => {
    const shallowWrapper = shallow(<TextInput {...baseProps} />);
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

  it('should pass on startAdornment when provided with one', () => {
    const props: TextInputProps = {
      ...baseProps,
      startAdornmentIcon: <Search />
    };
    const shallowWrapper = shallow(<TextInput {...props} />);
    expect((shallowWrapper.find('WithStyles(ForwardRef(Input))').props() as any).startAdornment).toBeTruthy();
  });

  it('should call onChange handler when input changes', () => {
    const props: TextInputProps = {
      ...baseProps,
      startAdornmentIcon: <Search />
    };
    const shallowWrapper = shallow(<TextInput {...props} />);
    shallowWrapper.find('WithStyles(ForwardRef(FormControl))').simulate('change');
    expect((props.onChange as any).mock.calls.length).toBe(1);
  });
});
