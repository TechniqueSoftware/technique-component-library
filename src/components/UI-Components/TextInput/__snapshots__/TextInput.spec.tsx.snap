// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should render with minimal props 1`] = `
"<WithStyles(ForwardRef(FormControl)) fullWidth={true} classes={{...}} value=\\"mockValue\\" onChange={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }}>
  <WithStyles(ForwardRef(Input)) inputProps={{...}} placeholder=\\"mockPlaceHolder\\" classes={{...}} fullWidth={true} autoFocus={false} startAdornment={{...}} />
</WithStyles(ForwardRef(FormControl))>"
`;
